### DISCUSSION TITLE
Reverse proxy with Nginx Proxy Manager

### DISCUSSION THREAD

**Discussion by gabviv73** [Mailu/Mailu#3658] [User support]
I'm getting mad trying to figure out how to configure mailu behind my existing nginx-proxy-manager reverse proxy.
It only works if I set TLS_FLAVOR=notls, but in this way, no imaps and pops are enabled.
If I use any other value for TLS_FLAVOUR, accessing web gui gives me too_many_redirects error.
Suggestions ?

**Comment by ksylvan** [comment]
You might be able to do some variant of the idea in this PR: https://github.com/Mailu/Mailu/pull/3684

Let imaps, pops, etc. be handled by MailU and deal with the web-based connections in your reverse proxy.

**Comment by leolivier** [comment]
Hi  
I have exactly the same issue as @gabviv73 and I don't get how @ksylvan's suggestion can help.
Nginx proxy manager provides SSL certificates from letsencrypt in its own mounted directory so I used `TLS_FLAVOR=cert`
and the following instructions in my docker-compose file in the front definition:
```
    volumes:
        - ../nginx-network-manager/letsencrypt/live/npm-6/fullchain.pem:/certs/cert.pem
        - ../nginx-network-manager/letsencrypt/live/npm-6/privkey.pem:/certs/key.pem
 ```
The container start very well and does not show any error in the logs but when I try to open https://mail.{mydomain} I get a bunch of 301 coming directly from the container (I see them in the container logs)

When I try to use `TLS_FLAVOR=mail-letsencrypt` and mount ./certs on /certs, and let Mailu deal with the certificates, I don't have these 301, but I get an error that the front can't reach the .well-known resources:
```
CRITICAL:root:Can't reach http://mail.{mydomain}/.well-known/acme-challenge/testing!, please ensure it's fixed or change the TLS_FLAVOR.
```

**Comment by leolivier** [comment]
I was finally able to get it working!
This is only a matter of configuration on NPM. What I did (as far as I can remember :) :
* I removed my existing proxy host and SSL certificates for mailu on NPM
* I created a new proxy host with an advanced block like this:
![image](https://github.com/user-attachments/assets/d4804b64-4a3a-437b-9d56-82209782a0c6)
and no SSL cert. 
* Then I started MailU (only the front normally is enough). This should allow Mailu to create its new cert with Let's Encrypt
* Then I created a custom certificate on NPM: 
![image](https://github.com/user-attachments/assets/1fa94328-0526-435a-9238-7de9dabc57fc)
![image](https://github.com/user-attachments/assets/94fdc782-dc1a-49f1-b26a-0edd129fc936)
You need to upload there the privkey.pem file generated by Mailu in previous steps in the certificate key field and the fullchain.pem in the certificate field. I also uploaded there the chain.pem in the intermediate field but I think it's not used (I can't retrieve it in the NPM directory afterwards)
* Then click Save and the certificate for MailU will now appear in the list of certificates.
* You can now update your proxy host and use the new certificate.

This has a major drawback indeed: you need to upload the certs again and again each time MailU renew the certificate so my next task is to update the certs automatically: using a cron job, I will copy the Mailu certs to the NPM directory each day or week (What I did through the NPM UI above can probably be done easily by copying the certs on the command line, I didn't check yet)

Hope this helps

**Reply by ZoranVelinov** [reply]
Hello,

Can you tell me what you write in Details?

<img width="501" height="549" alt="{57BDAA44-71C8-4C3F-B5BB-11D8E6B6BCF0}" src="https://github.com/user-attachments/assets/82001fbe-6f6e-4364-8072-2c4fad51b841" />


And have you find better solution than  cron job for certs?

Thank you in advance

**Reply by leolivier** [reply]
Hi @ZoranVelinov 
actually, I changed my setup to use cloudflare as first proxy layer. I'm still using NPM as a reverse proxy but in a different way...

**Comment by leolivier** [comment]
I wrote this very simple script to update the certs in NPM when MailU regenerates them:
```
#!/bin/bash

# update these 2 directories and the certificate number in npm for your mailu proxy host
mailu_root='/docker/mailu'
npm_root='/docker/nginx-proxy-manager'
npm_number=17

mailu_certs_dir="$mailu_root/certs/letsencrypt/live"
npm_certs_dir="$npm_root/data/custom_ssl/npm-${npm_number}"
sudo cp $mailu_certs_dir/mailu/privkey.pem $npm_certs_dir/privkey.pem
sudo cp $mailu_certs_dir/mailu-ecdsa/fullchain.pem $npm_certs_dir/fullchain.pem
```

You can then cron them daily

**Comment by nextgens** [comment]
Apparently you can override parts of nginx config with it https://develop.nginxproxymanager.com/advanced-config/#custom-nginx-configurations ...

It would be much better to document how to "stream_tcp" proxy to Mailu. That way no SSL termination would happen there at all.

**Reply by leolivier** [reply]
You're probably right @nextgens, but I have no idea how to do that. I'll look at that next weekend
